### PR TITLE
Add copy/paste functionality to edit.js

### PR DESCRIPTION
**Pull request by huffman**
https://highfidelity.manuscript.com/f/cases/19339/

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14138/HighFidelity-Beta-PR14235-4a9edf0.apk

**Comment by huffman** [comment]
# Test Plan

  1. Go to your domain
  1. Enable create
  1. Create a sphere
  1. Select the sphere
  1. Press ctrl-c to copy the selected entity
  1. Turn a small amount to the riht
  1. Press ctrl-v to paste the copied entity. Confirm that the sphere you copied has been duplicated
  1. Change the color of the new sphere to blue
  1. Back up a little bit.
  1. Select both spheres
  1. Press ctrl-c
  1. Press ctrl-v. Confirm that both spheres have been duplicated, and are the correct color (1 red, 1 blue)
  1. Make the red sphere the parent of the blue sphere
  1. Select only the red sphere
  1. Press ctrl-c
  1. Press ctrl-v. Confirm that both spheres have been duplicated (the parent and the child).
  1. Try using undo and redo and confirm that both work correctly with pasted entities

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14138/HighFidelity-Beta-PR14235-4a9edf0.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14138/HighFidelity-Beta-Interface-PR14235-4a9edf0.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14138/HighFidelity-Beta-PR14235-4a9edf0.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14138/HighFidelity-Beta-Interface-PR14235-4a9edf0.dmg

**Comment by shanzzam** [comment]
Is this a late-breaking fix for a bug? Otherwise, I can add it in 76.

**Comment by thoys** [comment]
This is planned work for 75 and is required for one of my PRs

File: scripts/system/edit.js [hunk 1]
```diff
@@ -2559,6 +2758,30 @@ var PopupMenu = function () {
...
+mapping.from([Controller.Hardware.Keyboard.Backspace]).when([Controller.Hardware.Application.PlatformMac]).to(deleteKey);
+mapping.from([Controller.Hardware.Keyboard.D]).when([Controller.Hardware.Keyboard.Control]).to(deselectKey);
+mapping.from([Controller.Hardware.Keyboard.T]).to(toggleKey);
+mapping.from([Controller.Hardware.Keyboard.F]).to(focusKey);
+mapping.from([Controller.Hardware.Keyboard.G]).to(gridKey);
+mapping.from([Controller.Hardware.Keyboard.C]).when([Controller.Hardware.Keyboard.Control]).to(copyKey);
+mapping.from([Controller.Hardware.Keyboard.V]).when([Controller.Hardware.Keyboard.Control]).to(pasteKey);
+
+function copyKey(value) {
+    console.log("Copy", value);
```
Comment (thoys) [inline scripts/system/edit.js]: Wouldn't you want to make this more descriptive if you keep this here?

File: scripts/system/edit.js [hunk 2]
```diff
@@ -2390,6 +2382,213 @@ var PropertiesTool = function (opts) {
...
+        if (avatar.getJointIndex(grabJointNames[i]) === properties.parentJointIndex) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+
+// entityProperites - array of entity property objects
```
Comment (thoys) [inline scripts/system/edit.js]: typo in entityProperites 

could use jsdoc instead e.g.
```
/**
 * Function description here
 * @param entityProperties {array} - Array of entity property objects
 */
```

File: scripts/system/edit.js [hunk 3]
```diff
@@ -2390,6 +2382,213 @@ var PropertiesTool = function (opts) {
...
+
+function pasteEntities() {
+    var dims = entityClipboard.dimensions;
+    var maxDim = Math.max(dims.x, dims.y, dims.z);
+    var pastePosition = getPositionToCreateEntity(maxDim);
+    var deltaPosition = Vec3.subtract(pastePosition, entityClipboard.position);
+
+    var copiedProperties = []
+    var ids = [];
+    entityClipboard.entities.forEach(function(origproperties) {
```
Comment (thoys) [inline scripts/system/edit.js]: please use camelcase originalProperties would be better to write here.

File: scripts/system/edit.js [hunk 4]
```diff
@@ -2390,6 +2382,213 @@ var PropertiesTool = function (opts) {
...
+        };
+    }
+}
+
+function deepCopy(v) {
+    return JSON.parse(JSON.stringify(v));
+}
+
+function pasteEntities() {
+    var dims = entityClipboard.dimensions;
```
Comment (thoys) [inline scripts/system/edit.js]: it would be more clear to name this variable `dimensions`

File: scripts/system/edit.js [hunk 5]
```diff
@@ -2390,6 +2382,213 @@ var PropertiesTool = function (opts) {
...
+    var maxDim = Math.max(dims.x, dims.y, dims.z);
```
Comment (thoys) [inline scripts/system/edit.js]: maxDimension would be more appropriate here.

File: scripts/system/edit.js [hunk 6]
```diff
@@ -2390,6 +2382,213 @@ var PropertiesTool = function (opts) {
...
+    var ids = [];
+    entityClipboard.entities.forEach(function(origproperties) {
+        var properties = deepCopy(origproperties);
+        if (properties.root) {
+            properties.position = Vec3.sum(properties.position, deltaPosition);
+            delete properties.localPosition;
+        } else {
+            delete properties.position;
+        }
+        //entityProperties[properties.id] = properties;
```
Comment (thoys) [inline scripts/system/edit.js]: please remove commented out code

File: interface/resources/controllers/keyboardMouse.json [hunk 7]
```diff
@@ -133,7 +133,7 @@
         { "from": "Keyboard.W", "when": "!Keyboard.Control", "to": "Actions.LONGITUDINAL_FORWARD" },
         { "from": "Keyboard.S", "when": "!Keyboard.Control", "to": "Actions.LONGITUDINAL_BACKWARD" },
         { "from": "Keyboard.Shift", "when": ["!Keyboard.Left", "!Keyboard.Right"], "to": "Actions.SPRINT" },
-        { "from": "Keyboard.C", "to": "Actions.VERTICAL_DOWN" },
+        { "from": "Keyboard.C", "when": "!Keyboard.Control", "to": "Actions.VERTICAL_DOWN" },
```
Comment (thoys) [inline interface/resources/controllers/keyboardMouse.json]: Nice, that would have been awkward while copying.

File: scripts/system/edit.js [hunk 8]
```diff
@@ -2390,6 +2382,213 @@ var PropertiesTool = function (opts) {
...
+        if (entityList.indexOf(childID) < 0) {
+            entityList.push(childID);
+        }
+        addChildrenEntities(childID, entityList);
+    }
+}
+
+// Return true if the given entity with `properties` is being grabbed by an avatar.
+// This is mostly a heuristic - there is no perfect way to know if an entity is being
+// grabbed.
```
Comment (thoys) [inline scripts/system/edit.js]: perhaps use jsdoc here instead?

File: scripts/system/edit.js [hunk 9]
```diff
@@ -2390,6 +2382,213 @@ var PropertiesTool = function (opts) {
...
+// entityProperites - array of entity property objects
+function createEntities(entityProperties) {
+    var entitiesToCreate = [];
+    var createdEntityIDs = [];
+    var createdChildrenWithOldParents = [];
+    var originalEntityToNewEntityID = [];
+
+    SelectionManager.saveProperties();
+
+    for (var i = 0, len = entityProperties.length; i < len; ++i) {
```
Comment (thoys) [inline scripts/system/edit.js]: please don't abbreviate where clear english could be used.

Comment (thoys) [reply scripts/system/edit.js]: sorry, didn't note I was talking about `len` here :)

**Comment by huffman** [comment]
@shanzzam John updated it for me, thanks for checking!

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14161/HighFidelity-Beta-PR14235-7d695c7.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14161/HighFidelity-Beta-PR14235-7d695c7.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14161/HighFidelity-Beta-Interface-PR14235-7d695c7.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14170/HighFidelity-Beta-PR14235-c18ee42.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14170/HighFidelity-Beta-Interface-PR14235-c18ee42.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14170/HighFidelity-Beta-PR14235-c18ee42.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14170/HighFidelity-Beta-Interface-PR14235-c18ee42.exe

**Comment by JeffClinton** [comment]
Build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14192/HighFidelity-Beta-PR14235-c18ee42.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14192/HighFidelity-Beta-PR14235-c18ee42.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14192/HighFidelity-Beta-Interface-PR14235-c18ee42.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14192/HighFidelity-Beta-PR14235-c18ee42.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14192/HighFidelity-Beta-Interface-PR14235-c18ee42.dmg

**Comment by JeffClinton** [comment]
Currently testing

File: scripts/system/edit.js [hunk 10]
```diff
@@ -2390,6 +2382,220 @@ var PropertiesTool = function (opts) {
...
+    function undo(copiedProperties) {
+        for (var i = 0; i < copiedProperties.length; ++i) {
+            Entities.deleteEntity(copiedProperties[i].id);
+        }
+        SelectionManager.setSelections(currentSelections);
+    }
+
+    redo(copiedProperties);
+    undoHistory.pushCommand(undo, copiedProperties, redo, copiedProperties);
+}
```
Comment (thoys) [inline scripts/system/edit.js]: lets move the copy, cut and paste functions into the selection manager

**Comment by JeffClinton** [comment]
NOTE: The fail issue noted above is present in master, so it should not have failed this PR.  However, changes have been made so this should be retested.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14259/HighFidelity-Beta-PR14235-573df2d.apk

**Comment by QliCknox** [comment]
Currently Testing

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14259/HighFidelity-Beta-PR14235-573df2d.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14259/HighFidelity-Beta-Interface-PR14235-573df2d.exe

**Comment by qligmaloney** [comment]
Build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14294/HighFidelity-Beta-PR14235-573df2d.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14294/HighFidelity-Beta-PR14235-573df2d.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14294/HighFidelity-Beta-Interface-PR14235-573df2d.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14294/HighFidelity-Beta-PR14235-573df2d.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14294/HighFidelity-Beta-Interface-PR14235-573df2d.exe

**Comment by QliCknox** [comment]
Currently Testing

**Comment by jakram** [comment]
Currently testing on PC/HMD

**Comment by QliCknox** [comment]
Pass @conklin94122 

Copy pasting work as the test plan intended, redo/undo functions with parenting and the copy/paste functionality

**Pull request merged**